Add FluidSynth tuning functions

forrcaho opened the pull request:
Here's the big FluidSynth tuning patch I promised.

I ran into an ugly problem of Chuck_Array4 containing a vector of 8-byte values on my 64-bit linux system, leading me to copy array elements (search "ugly hack" in the code). I hope there's actually a better way to handle that, but my code does work.

The new tuning functions are all documented with examples in a new fluidsynth-tuning.ck file.

forrcaho commented:
Have you had the chance to look at this yet?

lathertonj commented:
Sorry about the wait! Yup, Chuck_Array4 is for storing CKINTs which have 8 bytes when running in 64-bit mode. Your hack looks pretty reasonable, though depending on compiler warnings you might want to explicitly cast down to (int) when you're doing the copying?

I don't remember being able to test FluidSynth in the past (maybe it requires a separate install?), but I trust that everything works otherwise.

forrcaho commented:
Ok, added the suggested cast (even though it wasn't necessary in my env) and modified the ugly hack comment a bit.

The fluidsynth-tuning.ck file exercises all of the methods and that's what I'm using as my test.

FTR, I'm on Debian "unstable" for amd64 with g++ 8.2.0.

spencersalazar commented on FluidSynth/FluidSynth.cpp:
```diff
@@ -147,7 +226,43 @@ CK_DLL_QUERY( fluidsynth )
     QUERY->add_arg(QUERY, "int", "bankNum");
     QUERY->add_arg(QUERY, "int", "chan");
 
-    fluidsynth_data_offset = QUERY->add_mvar(QUERY, "int", "@f_data", false);
```
Im confused why this line was removed / how this chugin works without this line? This is needed to allocate a space in the Chuck_Object for the corresponding instance of the FluidSynth class.

spencersalazar commented:
Thanks Forrest. I don't find your solution particularly "hacky" considering the circumstances so no worries there. As Jack mentioned Chuck_Array4 is used for storing "int-sized" things like ints and pointers, which historically in ChucK's early development were 4 bytes on 32 bit systems. 

Before merging Im just curious about why the member variable for storing the FluidSynth class instance was removed?

forrcaho commented:
Removing that `fluidsynth_data_offset` line was definitely an oversight; thanks for catching it. I've re-inserted it (at the end of the command definitions, which matches other chugins I checked). Although the code without that line worked for me, I guess I may just have been lucky it wasn't being used for something else.

forrcaho commented:
So this patch is good to go now, right?

forrcaho commented:
?

spencersalazar commented:
All good, thanks 👍
